t/helper/test-gvfs-protocol.c

```diff
@@ -0,0 +1,1191 @@
+
+	return my_send_loose_object(&oi, &oid, 1);
+}
+
+static int do__gvfs_objects__post(struct my_request_data *req_data)
+{
+	return -1;
+}
+	
+#if 0
+/*
+ * Parse line 0 of the HTTP request.  We expect something of the form:
+ *
+ *    GET [<uri_base>]/<gvfs_api>[/<args>] HTTP/1.1
+ *    GET [<uri_base>]/<gvfs_api>[?<args>] HTTP/1.1
+ *
+ * where:
+ *    <uri_base> is the URI (excluding the protocol and hostname) used
+ *               by the client.
+ *    <gvfs_api> is the REST API named in the GVFS Protocol.  This is
+ *               of the form "gvfs/objects".
+ *    <args>     is any args to the above, such as "<oid>".
+ *
+ * For example, when the cache-server or main Git server is set to
+ * something like "http://localhost:9999/", we get:
+ *
+ *    GET /gvfs/objects/aaaaaaaaaabbbbbbbbbbccccccccccdddddddddd HTTP/1.1
+ *    POST /gvfs/objects HTTP/1.1
+ *
+ * We can use the base part of the URI to support both types of servers
+ * on one port.  For example, with "http://localhost:9999/repos/" and
+ * "http://localhost:9999/cache/", we get:
+ *
+ *    GET /repos/gvfs/objects/aaaaaaaaaabbbbbbbbbbccccccccccdddddddddd HTTP/1.1
+ *    GET /cache/gvfs/objects/aaaaaaaaaabbbbbbbbbbccccccccccdddddddddd HTTP/1.1
+ *
+ * Or we can let <uri_base> be a series of key/value pairs to drive
+ * different configurations or unit tests.  For example:
+ *
+ *    GET /k1/v1/k2/v2/gvfs/objects/aaaaaaaaaabbbbbbbbbbccccccccccdddddddddd HTTP/1.1
+ * 
+ * We only expect a limited set of URLs for testing purposes, so we
+ * do not bother with full URL escape sequence decoding here.
+ */    
+#endif
+
+/*
+ * Read the HTTP request up to the start of the optional message-body.
+ * We do this byte-by-byte because we have keep-alive turned on and
+ * cannot rely on an EOF.
+ *
+ * https://tools.ietf.org/html/rfc7230
+ * https://github.com/microsoft/VFSForGit/blob/master/Protocol.md
+ *
+ * Returns 1 on EOF
+ * Returns 0 when we have a valid request
+ * Returns -1 when we have a parser error.
+ *
+ * We cannot call die() here because our caller needs to properly
+ * respond to the client and/or close the socket before this
+ * child exits so that the client doesn't get a connection reset
+ * by peer error.
+ */
+static int my_request_data__read(struct my_request_data *req_data, int fd)
+{
+	struct strbuf h = STRBUF_INIT;
+	int nr_start_line_fields;
+	const char *uri_target;
+	const char *http_version;
+	const char *gvfs;
+
+	/*
+	 * Read line 0 of the request and split it into component parts:
+	 *
+	 *    <method> SP <uri-target> SP <HTTP-version> CRLF
+	 *
+	 */    
+	if (strbuf_getwholeline_fd(&req_data->start_line, fd, '\n') == EOF)
+		return 1;
+
+	if (string_list_has_string(&mayhem_list, "close_read")) {
+		logmayhem("mayhem: close_read");
+		return -1;
+	}
+
+	strbuf_trim_trailing_newline(&req_data->start_line);
+
+	nr_start_line_fields = string_list_split(&req_data->start_line_fields,
+						 req_data->start_line.buf,
+						 ' ', -1);
```

Comment:
Isn't a full `string_list_split()` a bit heavy-handed here? We can easily use `strbuf_strip_suffix(&req->start_line, "HTTP/1.1")` here, followed by `skip_prefix(req->start_line.buf, "GET ", &uri_target)` (checking the return values of both calls, of course). If you want, you can then verify `if (strchr(uri_target, ' ')) { logerror("too many GET parameters: '%s'", uri_target); return -1; }`

Reply:
Yeah, an earlier version did something like that, but it was starting to venture off into the weeds -- lets check for GET and POST (and maybe later HEAD) and HTTP/1.1 and HTTP/2 and etc.  And then I wanted to have the URI in a field by itself to cut apart the endpoint fields and etc.  So it was (conceptually) simpler to just split it and move on.